peps/pep-0734.rst

```diff
@@ -0,0 +1,396 @@
+API is not intended for public use and thus not part of this proposal.
+
+We also expect that an ``InterpreterPoolExecutor`` will be added to the
+``concurrent.futures`` module, but that is outside the scope of this PEP.
+
+API: Using Interpreters
+-----------------------
+
+The module's top-level API for managing interpreters looks like this:
+
++----------------------------------+----------------------------------------------+
+| signature                        | description                                  |
++==================================+==============================================+
+| ``list_all() -> [Interpreter]``  | Get all existing interpreters.               |
++----------------------------------+----------------------------------------------+
+| ``get_current() -> Interpreter`` | Get the currently running interpreter.       |
++----------------------------------+----------------------------------------------+
+| ``create() -> Interpreter``      | Initialize a new (idle) Python interpreter.  |
++----------------------------------+----------------------------------------------+
+
+Each interpreter object:
+
++----------------------------------+------------------------------------------------+
+| signature                        | description                                    |
++==================================+================================================+
+| ``class Interpreter``            | A single interpreter.                          |
++----------------------------------+------------------------------------------------+
+| ``.id``                          | The interpreter's ID (read-only).              |
++----------------------------------+------------------------------------------------+
+| ``.is_running() -> bool``        | Is the interpreter currently executing code?   |
++----------------------------------+------------------------------------------------+
+| ``.set_main_attrs(**kwargs)``    | Bind objects in ``__main__``.                  |
++----------------------------------+------------------------------------------------+
+| ``.exec(code, /)``               | | Run the given source code in the interpreter |
+|                                  | | (in the current thread).                     |
++----------------------------------+------------------------------------------------+
+
+Additional details:
+
+* Every ``Interpreter`` instance adds a reference to the corresponding
+  interpreter, which is released when the ``Interpreter`` instance is
+  deleted.  When there are no more such references to an interpreter,
+  it gets finalized.  Thus no interpreters created through
+  ``interpreters.create()`` will leak.
+
+* ``Interpreter.is_running()`` refers only to if there is a thread
+  running a script (code) in the interpreter's ``__main__`` module.
+  That basically means whether or not ``Interpreter.exec()`` is running
+  in some thread.  Code running in sub-threads is ignored.
```

Comment:
The main practical value I see is being able to query before calling `interp.exec()`, to avoid getting an exception if it's already running.  Otherwise it's similar to `Thread.is_alive()` and would have similar utility.  I also imagine it could be helpful for people poking around and exploring how interpreters work.